3단계 - 자동차 경주

yamsroun opened the pull request:
"3단계 - 자동차 경주"의 코드 리뷰 요청 드립니다.
(`StringCalculator`는 심심해서(?) 수정을 했었는데, 무시해 주세요.)

yamsroun committed:
2단계 > 문자열 덧셈 계산기 (리뷰 반영 + 리팩토링)
- 커스텀 구분자 패턴을 처리하는 클래스 분리

yamsroun committed:
Merge branch 'step2' into yamsroun

yamsroun committed:
docs: 기능 목록 작성

yamsroun committed:
feat: 기능 목록 구현

nooose commented on src/main/java/racingcar/ui/RacingConsoleUi.java:
```diff
@@ -0,0 +1,48 @@
+public class RacingConsoleUi implements RacingUi {
+
+    public static final String CAR_START_UNIT = "|";
+    public static final String CAR_POSITION_UNIT = "-";
```
UI에서만 사용하는 상수같아서 private 접근제한자가 더 좋을 것 같아요

nooose commented on src/main/java/racingcar/ui/RacingConsoleUi.java:
```diff
@@ -0,0 +1,48 @@
+            print(CAR_START_UNIT);
+            for (int i = 0; i < position; i++) {
+                print(CAR_POSITION_UNIT);
+            }
```
사실 표준출력을 거의 사용할 일은 없겠지만
자동차 1대가 100칸 이동했다면 출력요청이 그만큼 많아질 것 같아요
자동차 거리만큼 텍스트그림(?)을 만들고 한대한대씩 출력하는 게 더 좋을 것 같습니다.

nooose commented on src/main/java/racingcar/racing/RacingGame.java:
```diff
@@ -0,0 +1,50 @@
+        RacingUi racingUi = new RacingConsoleUi();
+        RacingPreference preference = racingUi.inputPreference();
+        RacingGame racingGame = new RacingGame(preference, racingUi, new Random());
+        racingGame.start();
```
```suggestion
        RacingUi racingUi = new RacingConsoleUi();
        RacingPreference preference = racingUi.inputPreference();
        RacingGame racingGame = new RacingGame(preference, new Random());
        Result result = racingGame.start();
        racingUi.출력(result); 
```
입력은 RacingGame 밖에서 받는데 출력은 RacingGame 내부에서 사용하는 것이 약간 어색하네요
위와 같이 수정된다면 입출력과 비즈니스 로직을 완전히 분리시킬 수 있을 것 같아요
분리된다면 RacingGame 테스트도 쉽게할 수 있어보입니다.

_어느 시점에 출력할 것인지에 대한 제약은 없다._ 조건이 있지만
팀장님은 남달라서 리뷰드렸습니다.

nooose commented on src/main/java/racingcar/racing/RacingPreference.java:
```diff
@@ -0,0 +1,7 @@
+package racingcar.racing;
+
+public record RacingPreference(
+    int carCount,
+    int moveCount
+) {
```
👍 DTO 좋네용

nooose commented on src/test/java/racingcar/mock/Always4OrMoreRandom.java:
```diff
@@ -0,0 +1,11 @@
+
+import java.util.Random;
+
+public class Always4OrMoreRandom extends Random {
```
너무 좋습니다.
테스트에서 랜덤 제어가 가능하겠네요 👍

nooose commented on src/main/java/racingcar/entry/Car.java:
```diff
@@ -0,0 +1,29 @@
+
+public class Car {
+
+    private final Random random;
```
지금은 Random을 받아도 괜찮은데
요구사항이 변경되어 다른 환경에서는 범위가 넓은 랜덤값 또는 외부 요인(날씨, API)으로 전진이 판단된다면 Car 도메인이 많이 변경될 것 같아요
전략 패턴을 사용해서 리팩토링 해보면 어떨까요?

```java
interface MovePolicy {
    boolean isMovable();
}
```

```suggestion
    private final MovePolicy policy;
```

nooose commented on src/test/java/racingcar/mock/AlwaysLessThan4Random.java:
```diff
@@ -0,0 +1,11 @@
+
+import java.util.Random;
+
+public class AlwaysLessThan4Random extends Random {
```
👍

nooose commented on src/main/java/racingcar/racing/RacingEntries.java:
```diff
@@ -0,0 +1,30 @@
+    }
+
+    public List<Car> getEntries() {
+        return Collections.unmodifiableList(cars);
```
리스트는 불변인데 리스트 내부 자동차는 얕은 복사 상태이기 때문에
만약 외부에서 자동차를 꺼내고 이동시킨다면 일급 컬렉션 내 값들이 변경되는 문제가 발생할 것 같아요
깊은 복사를 하면 문제가 해결될 것 같습니다.

nooose commented on src/main/java/racingcar/entry/Car.java:
```diff
@@ -0,0 +1,29 @@
+
+    private final Random random;
+
+    private int position = 0;
```
증가만 해야하는 값이기 때문에 음수가 되지 않도록
원시타입 position을 객체로 래핑해보면 좋겠네요 🤗

nooose reviewed (changes requested):
코드가 매우 깔끔하네요 🧼
그래도 재요청 부탁드립니다. 

사정없이 해달라는 말에..

yamsroun committed:
refactor: 코드 리뷰 반영
- 직접적 `Random` 의존성 제거
- 이동 정책을 의미하는 `MovePolicy` 인터페이스 정의
- `MovePolicy`를 생성자 또는 전략 패턴으로 지정 가능
- `Car` 위치를 표현하는 객체 생성(`CarPosition`)
- `List` 조회 메서드에서 Deep Copy 반영
- `RacingGame`과 `RacingUi` 의존성 분리
- 게임 결과 출력 로직 개선
- 객체 복제 생성자 추가
- 내부에서만 사용하는 필드를 `private`으로 수정
- 테스트 코드 개선

yamsroun commented:
꼼꼼한 코드 리뷰 감사합니다. 😍😍😍
리뷰해주신 내용 모두 반영했습니다. 🙌🏾🙌🏾🙌🏾
추가로 리뷰할 부분 있으면 또 말씀해 주세요. 🙏🏾🙏🏾🙏🏾
덕분에 많이 배웠습니다. 🙇🏾‍♀️🙇🏾‍♂️🙇🏾

nooose commented on src/test/java/racingcar/entry/CarTest.java:
```diff
@@ -0,0 +1,64 @@
+    @RepeatedTest(10)
+    @DisplayName("Setter를 통해 MovePolicy를 설정하고, 랜덤값이 4 미만이면 전진하지 않는다.")
+    void notMove() {
+        MovePolicy movePolicy = new AlwaysNotMovePolicy();
+        Car car = new Car();
+        car.setMovePolicy(movePolicy);
+
+        int prevPosition = car.currentPosition();
+        car.move();
```
```java
        Car car = new Car();

        int prevPosition = car.currentPosition();
        car.move(() -> false);
```

세터 대신 메소드에 policy를 전달하는 방법도 있습니다. 
그냥 이거는 참고 ㅎㅎ

nooose commented on src/main/java/racingcar/entry/Car.java:
```diff
@@ -0,0 +1,42 @@
+        this.movePolicy = movePolicy;
+    }
+
+    public Car(Car car) {
+        this.movePolicy = car.movePolicy;
+        this.position = new CarPosition(car.position.current());
+    }
```
딥카피 용 생성자네요 👍

이 방법도 좋지만 Car 자체를 불변으로 만들면
리스트 게터로 내보낸 이후에 수정을 해도 이전 리스트 내 원소에 영향이 없어지기 때문에
딥카피를 하지않아도 됩니다. 스냅샷도 바로바로 기록 가능하겠네요

Car 가 아니라 position을 불변으로 만들어도 
Car도 자연스럽게 불변으로 따라가게 되어있습니다.

이번 리뷰건은 참고라서 수정하지 않으셔도됩니다. 😄

- 예시
```java
    public Car move() {
        if (canMove()) {
            return new Car(position.increase());
        }
    }
```

```java
    Car movedCar = car.move();
```

nooose commented on src/main/java/racingcar/racing/RacingGame.java:
```diff
@@ -0,0 +1,50 @@
+
+
+    public static void main(String[] args) {
+        RacingUi racingUi = new ConsoleRacingUi();
+        RacingPreference preference = racingUi.inputPreference();
+        RacingGame racingGame = new RacingGame(preference, new RandomMovePolicy());
+        racingUi.showResult(racingGame.race());
```
👍

nooose reviewed (approved):
저도 많이 배웠습니다.
아주 제대로 반영되었네요
머지하겠습니다. 👍👍

yamsroun commented:
감사합니다~ 😍
